src/dialogue_helpers.h

```diff
@@ -147,13 +148,24 @@ std::string read_var_value( const var_info &info, const dialogue &d );
 
 var_info process_variable( const std::string &type );
 
-
-struct str_or_var {
-    std::optional<std::string> str_val;
+template<class T>
+struct abstract_str_or_var {
+    std::optional<T> str_val;
     std::optional<var_info> var_val;
-    std::optional<std::string> default_val;
+    std::optional<T> default_val;
     std::optional<std::function<std::string( const dialogue & )>> function;
-    std::string evaluate( dialogue const &d ) const;
+    virtual ~abstract_str_or_var() = default;
+    virtual std::string evaluate( dialogue const & ) const {
+        return std::string();
+    }
+};
+
+struct str_or_var : abstract_str_or_var<std::string> {
+    std::string evaluate( dialogue const &d ) const override;
+};
+
+struct translation_or_var : abstract_str_or_var<translation> {
+    std::string evaluate( dialogue const &d ) const override;
 };
```

Comment:
AFAICT `str_or_val` and `translation_or_val` are not used dynamically so you can probably define this class statically and define `str_or_var` and `translation_or_var` to be specializations of the template, like this:

```suggestion
    std::string evaluate( dialogue const &d ) const;
};

using str_or_var = abstract_str_or_var<std::string>;
using translation_or_var = abstract_str_or_var<translation>;
```